### DISCUSSION TITLE
jsonata return key value pairs

### DISCUSSION THREAD

**Discussion by yosiasz** [grafana/grafana-infinity-datasource#558] [Q&A] [answered]
Greetings,

How can I return the data from this jsonata

https://try.jsonata.org/7uPiakxSk

```
{
  "cost": 609.0549451402,
  "provider": "azurefirewall_vwan-hub"
}
```

in order to graph it in grafana when using this in UQL it errors.

```
parse-json
| jsonata "( $data :=$.properties.rows; $final := { "cost": $data[0][0], "provider": $data[0][1].$split("/")[8] } )"
```
Thanks!

**Comment by yesoreyeram** [comment, marked as answer]
Hi @yosiasz - sorry for the delay. Actually you don't need JSONata at all in this case with UQL. ( if you still want jsonata, provided the sample below )

* [UQL](https://play.grafana.org/explore?orgId=1&left=%7B%22datasource%22:%22Yzeq07F4k%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22datasource%22:%7B%22type%22:%22yesoreyeram-infinity-datasource%22,%22uid%22:%22Yzeq07F4k%22%7D,%22type%22:%22json%22,%22source%22:%22inline%22,%22format%22:%22table%22,%22url%22:%22https:%2F%2Fgithub.com%2Fyesoreyeram%2Fgrafana-infinity-datasource%2Fblob%2Fmain%2Ftestdata%2Fusers.json%22,%22url_options%22:%7B%22method%22:%22GET%22,%22data%22:%22%22%7D,%22root_selector%22:%22%22,%22columns%22:%5B%5D,%22filters%22:%5B%5D,%22global_query_id%22:%22%22,%22parser%22:%22uql%22,%22data%22:%22%7B%5Cn%20%20%5C%22eTag%5C%22:%20null,%5Cn%20%20%5C%22id%5C%22:%20%5C%22subscriptions%2FXXXXX%2Fproviders%2FMicrosoft.CostManagement%2Fquery%2Fff4cc502-7c90-4c35-bc67-b3b060fe3a8c%5C%22,%5Cn%20%20%5C%22location%5C%22:%20null,%5Cn%20%20%5C%22name%5C%22:%20%5C%22ff4cc502-7c90-4c35-bc67-b3b060fe3a8c%5C%22,%5Cn%20%20%5C%22properties%5C%22:%20%7B%5Cn%20%20%20%20%5C%22columns%5C%22:%20%5B%5Cn%20%20%20%20%20%20%7B%5Cn%20%20%20%20%20%20%20%20%5C%22name%5C%22:%20%5C%22PreTaxCost%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22type%5C%22:%20%5C%22Number%5C%22%5Cn%20%20%20%20%20%20%7D,%5Cn%20%20%20%20%20%20%7B%5Cn%20%20%20%20%20%20%20%20%5C%22name%5C%22:%20%5C%22ResourceId%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22type%5C%22:%20%5C%22String%5C%22%5Cn%20%20%20%20%20%20%7D,%5Cn%20%20%20%20%20%20%7B%5Cn%20%20%20%20%20%20%20%20%5C%22name%5C%22:%20%5C%22Currency%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22type%5C%22:%20%5C%22String%5C%22%5Cn%20%20%20%20%20%20%7D%5Cn%20%20%20%20%5D,%5Cn%20%20%20%20%5C%22nextLink%5C%22:%20null,%5Cn%20%20%20%20%5C%22rows%5C%22:%20%5B%5Cn%20%20%20%20%20%20%5B%5Cn%20%20%20%20%20%20%20%20609.05494514021,%5Cn%20%20%20%20%20%20%20%20%5C%22%2Fsubscriptions%2FXXX%2Fresourcegroups%2Fpaw%2Fproviders%2Fmicrosoft.network%2Fazurefirewalls%2Fazurefirewall_vwan-hub%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22USD%5C%22%5Cn%20%20%20%20%20%20%5D,%5Cn%20%20%20%20%20%20%5B%5Cn%20%20%20%20%20%20%20%20709.05494514021,%5Cn%20%20%20%20%20%20%20%20%5C%22%2Fsubscriptions%2FXXX%2Fresourcegroups%2Fpaw%2Fproviders%2Fmicrosoft.network%2Fazurefirewalls%2Fazurefirewall_vwan-hub2%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22USD%5C%22%5Cn%20%20%20%20%20%20%5D%5Cn%20%20%20%20%5D%5Cn%20%20%7D,%5Cn%20%20%5C%22sku%5C%22:%20null,%5Cn%20%20%5C%22type%5C%22:%20%5C%22Microsoft.CostManagement%2Fquery%5C%22%5Cn%7D%22,%22uql%22:%22parse-json%5Cn%7C%20scope%20%5C%22properties.rows%5C%22%5Cn%7C%20extend%20%5C%22providers%5C%22%3Dsplit%28%5C%221%5C%22,%27%2F%27%29%5Cn%7C%20extend%20%5C%22provider%5C%22%3D%5C%22providers%5B8%5D%5C%22%5Cn%7C%20extend%20%5C%22cost%5C%22%3D%5C%220%5C%22%5Cn%7C%20project%20%5C%22provider%5C%22,%5C%22cost%5C%22%22%7D%5D,%22range%22:%7B%22from%22:%22now-1h%22,%22to%22:%22now%22%7D%7D)

* [UQL With JSONatata](https://play.grafana.org/explore?orgId=1&left=%7B%22datasource%22:%22Yzeq07F4k%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22datasource%22:%7B%22type%22:%22yesoreyeram-infinity-datasource%22,%22uid%22:%22Yzeq07F4k%22%7D,%22type%22:%22json%22,%22source%22:%22inline%22,%22format%22:%22table%22,%22url%22:%22https:%2F%2Fgithub.com%2Fyesoreyeram%2Fgrafana-infinity-datasource%2Fblob%2Fmain%2Ftestdata%2Fusers.json%22,%22url_options%22:%7B%22method%22:%22GET%22,%22data%22:%22%22%7D,%22root_selector%22:%22%22,%22columns%22:%5B%5D,%22filters%22:%5B%5D,%22global_query_id%22:%22%22,%22parser%22:%22uql%22,%22data%22:%22%7B%5Cn%20%20%5C%22eTag%5C%22:%20null,%5Cn%20%20%5C%22id%5C%22:%20%5C%22subscriptions%2FXXXXX%2Fproviders%2FMicrosoft.CostManagement%2Fquery%2Fff4cc502-7c90-4c35-bc67-b3b060fe3a8c%5C%22,%5Cn%20%20%5C%22location%5C%22:%20null,%5Cn%20%20%5C%22name%5C%22:%20%5C%22ff4cc502-7c90-4c35-bc67-b3b060fe3a8c%5C%22,%5Cn%20%20%5C%22properties%5C%22:%20%7B%5Cn%20%20%20%20%5C%22columns%5C%22:%20%5B%5Cn%20%20%20%20%20%20%7B%5Cn%20%20%20%20%20%20%20%20%5C%22name%5C%22:%20%5C%22PreTaxCost%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22type%5C%22:%20%5C%22Number%5C%22%5Cn%20%20%20%20%20%20%7D,%5Cn%20%20%20%20%20%20%7B%5Cn%20%20%20%20%20%20%20%20%5C%22name%5C%22:%20%5C%22ResourceId%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22type%5C%22:%20%5C%22String%5C%22%5Cn%20%20%20%20%20%20%7D,%5Cn%20%20%20%20%20%20%7B%5Cn%20%20%20%20%20%20%20%20%5C%22name%5C%22:%20%5C%22Currency%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22type%5C%22:%20%5C%22String%5C%22%5Cn%20%20%20%20%20%20%7D%5Cn%20%20%20%20%5D,%5Cn%20%20%20%20%5C%22nextLink%5C%22:%20null,%5Cn%20%20%20%20%5C%22rows%5C%22:%20%5B%5Cn%20%20%20%20%20%20%5B%5Cn%20%20%20%20%20%20%20%20609.05494514021,%5Cn%20%20%20%20%20%20%20%20%5C%22%2Fsubscriptions%2FXXX%2Fresourcegroups%2Fpaw%2Fproviders%2Fmicrosoft.network%2Fazurefirewalls%2Fazurefirewall_vwan-hub%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22USD%5C%22%5Cn%20%20%20%20%20%20%5D,%5Cn%20%20%20%20%20%20%5B%5Cn%20%20%20%20%20%20%20%20709.05494514021,%5Cn%20%20%20%20%20%20%20%20%5C%22%2Fsubscriptions%2FXXX%2Fresourcegroups%2Fpaw%2Fproviders%2Fmicrosoft.network%2Fazurefirewalls%2Fazurefirewall_vwan-hub2%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22USD%5C%22%5Cn%20%20%20%20%20%20%5D%5Cn%20%20%20%20%5D%5Cn%20%20%7D,%5Cn%20%20%5C%22sku%5C%22:%20null,%5Cn%20%20%5C%22type%5C%22:%20%5C%22Microsoft.CostManagement%2Fquery%5C%22%5Cn%7D%22,%22uql%22:%22parse-json%5Cn%7C%20jsonata%20%5C%22%28%20$final%20:%3D%20$map%28$.properties.rows,function%28$v%29%20%7B%20return%20%7B%20%20%27provider%27%20:%20$split%28$v%5B1%5D,%27%2F%27%29%5B8%5D,%20%27cost%27%20:%20$v%5B0%5D%20%7D%20%7D%29%20%29%5C%22%22%7D%5D,%22range%22:%7B%22from%22:%22now-1h%22,%22to%22:%22now%22%7D%7D) and your JSONata  is [here](https://try.jsonata.org/Ba2t5wwMr).
* [Backend](https://play.grafana.org/explore?orgId=1&left=%7B%22datasource%22:%22Yzeq07F4k%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22datasource%22:%7B%22type%22:%22yesoreyeram-infinity-datasource%22,%22uid%22:%22Yzeq07F4k%22%7D,%22type%22:%22json%22,%22source%22:%22inline%22,%22format%22:%22table%22,%22url%22:%22https:%2F%2Fgithub.com%2Fyesoreyeram%2Fgrafana-infinity-datasource%2Fblob%2Fmain%2Ftestdata%2Fusers.json%22,%22url_options%22:%7B%22method%22:%22GET%22,%22data%22:%22%22%7D,%22root_selector%22:%22%28%20%5Cn%20%20%20%20$final%20:%3D%20$map%28$.properties.rows,function%28$v%29%20%7B%5Cn%20%20%20%20%20%20%20%20return%20%7B%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%5C%22cost%5C%22%20:%20$v%5B0%5D,%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%5C%22provider%5C%22%20:%20$split%28$v%5B1%5D,%27%2F%27%29%5B8%5D%5Cn%20%20%20%20%20%20%20%20%7D%5Cn%20%20%20%20%7D%29%5Cn%29%22,%22columns%22:%5B%5D,%22filters%22:%5B%5D,%22global_query_id%22:%22%22,%22parser%22:%22backend%22,%22data%22:%22%7B%5Cn%20%20%5C%22eTag%5C%22:%20null,%5Cn%20%20%5C%22id%5C%22:%20%5C%22subscriptions%2FXXXXX%2Fproviders%2FMicrosoft.CostManagement%2Fquery%2Fff4cc502-7c90-4c35-bc67-b3b060fe3a8c%5C%22,%5Cn%20%20%5C%22location%5C%22:%20null,%5Cn%20%20%5C%22name%5C%22:%20%5C%22ff4cc502-7c90-4c35-bc67-b3b060fe3a8c%5C%22,%5Cn%20%20%5C%22properties%5C%22:%20%7B%5Cn%20%20%20%20%5C%22columns%5C%22:%20%5B%5Cn%20%20%20%20%20%20%7B%5Cn%20%20%20%20%20%20%20%20%5C%22name%5C%22:%20%5C%22PreTaxCost%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22type%5C%22:%20%5C%22Number%5C%22%5Cn%20%20%20%20%20%20%7D,%5Cn%20%20%20%20%20%20%7B%5Cn%20%20%20%20%20%20%20%20%5C%22name%5C%22:%20%5C%22ResourceId%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22type%5C%22:%20%5C%22String%5C%22%5Cn%20%20%20%20%20%20%7D,%5Cn%20%20%20%20%20%20%7B%5Cn%20%20%20%20%20%20%20%20%5C%22name%5C%22:%20%5C%22Currency%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22type%5C%22:%20%5C%22String%5C%22%5Cn%20%20%20%20%20%20%7D%5Cn%20%20%20%20%5D,%5Cn%20%20%20%20%5C%22nextLink%5C%22:%20null,%5Cn%20%20%20%20%5C%22rows%5C%22:%20%5B%5Cn%20%20%20%20%20%20%5B%5Cn%20%20%20%20%20%20%20%20609.05494514021,%5Cn%20%20%20%20%20%20%20%20%5C%22%2Fsubscriptions%2FXXX%2Fresourcegroups%2Fpaw%2Fproviders%2Fmicrosoft.network%2Fazurefirewalls%2Fazurefirewall_vwan-hub%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22USD%5C%22%5Cn%20%20%20%20%20%20%5D,%5Cn%20%20%20%20%20%20%5B%5Cn%20%20%20%20%20%20%20%20709.05494514021,%5Cn%20%20%20%20%20%20%20%20%5C%22%2Fsubscriptions%2FXXX%2Fresourcegroups%2Fpaw%2Fproviders%2Fmicrosoft.network%2Fazurefirewalls%2Fazurefirewall_vwan-hub2%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22USD%5C%22%5Cn%20%20%20%20%20%20%5D%5Cn%20%20%20%20%5D%5Cn%20%20%7D,%5Cn%20%20%5C%22sku%5C%22:%20null,%5Cn%20%20%5C%22type%5C%22:%20%5C%22Microsoft.CostManagement%2Fquery%5C%22%5Cn%7D%22%7D%5D,%22range%22:%7B%22from%22:%22now-1h%22,%22to%22:%22now%22%7D%7D) ( more power with alerts, query caching, recorded queries, public dashboards etc )

```js
( 
    $final := $map($.properties.rows,function($v) {
        return {
            "cost" : $v[0],
            "provider" : $split($v[1],'/')[8]
        }
    })
)
```

**Reply by yesoreyeram** [reply]
BTW, i think the error is because, you are using double quotes within your jsonata instead of single quotes which breaks in UQL. The syntax of UQL+JSONata is 

```kusto
parse-json
| jsonata "some_jsonata_goes_here_with_all_double_quotes_replaced_with_single_quotes_in_a_single_line"
```

I know it is pain to write everything in a single line and double quotes replaced with single quotes.. But I will find a way to support multiline JSONata when I have time. You can also use backend parser which overcomes this already.
